onnxruntime/core/providers/snpe/SnpeLib.cc

```diff
@@ -0,0 +1,446 @@
+#include "core/providers/snpe/SnpeLib.h"
+#include <iostream>
+#include <unordered_map>
+#include "core/common/common.h"
+#include "DlSystem/IUserBufferFactory.hpp"
+
+namespace onnxruntime {
+namespace contrib {
+namespace snpe {
+
+size_t CalcSizeFromDims(const zdl::DlSystem::Dimension* dims, size_t rank, size_t elementSize) {
+  if (rank == 0) return 0;
+  size_t size = elementSize;
+  while (rank--) {
+    size *= *dims;
+    dims++;
+  }
+  return size;
+}
+
+bool SnpeLib::SetupUserBufferAttribute(const std::string& name) {
+  auto buffer_attributes = snpe_->getInputOutputBufferAttributes(name.c_str());
```

Comment:
Not sure how big the attributed are. If you don't have to modify the local copy can receive it in a ref. ```auto& buffer_attributes```

Reply:
> Not sure how big the attributed are. If you don't have to modify the local copy can receive it in a ref. `auto& buffer_attributes`

the return type is zdl::DlSystem::Optional, not able to use reference.